src/python/pants/rules/core/test.py

```diff
@@ -66,42 +64,29 @@ class TestDebugRequest:
 
 
 @union
-class TestTarget:
-    """A union for registration of a testable target type.
+@dataclass(frozen=True)  # type: ignore[misc]   # https://github.com/python/mypy/issues/5374
+class TestRunner(ABC):
```

Comment:
Relative to calling this `TestTarget`, calling this `TestRunner` feels like a bit of a misnomer. This dataclass does not implement a test runner... rather, it is a unique class for a particular test type that a rule can consume. Giving it the `adaptor_with_origin` field and continuing to call it `TestTarget` might be clearer.

Not critical as I think that we can continue to iterate on it. But we should continue to watch out for "is-a" (or find some new way to describe unions.)

Reply:
True. This also applies to `Linter` and `Formatter`, which have a single field for `adaptors_with_origins`. I'll think if there's any more descriptive name..

It is accurate that we're describing the tools now, e.g. `BanditLinter` and `PytestTestRunner`. But, we're describing a specific instance of it, rather than a generic thing like a subsystem. Maybe `BanditLinterInstance`? I'll gladly trade a little more verbosity for something more descriptive.